### ISSUE TITLE
Problem with C handling of ion_int with large integer values. 

### ISSUE THREAD

**Issue by zabele** [amazon-ion/ion-c#327]
I was having memory corruption problems with big integers using ion_int types when the integers are larger than will fit in an int64_t. (e.g., 23456789012345678901).
I then went back and compiled the C example from Reading numeric types from the Cookbook verbatim and it core dumps when it gets to the ion_int part.
Seems to be a bug here somewhere.

**Comment by nirosys** [comment]
Hello zabele! Thank you for posting this issue, it looks like our cookbook entry has a bug.

I first duplicated a segmentation fault with the cookbook code:
```
$ src/IonCInt
ion_decimal: 1.23456
     double: 1234500.000000
        int: 123456
fish: Job 1, 'src/IonCInt' terminated by signal SIGSEGV (Address boundary error)
```

Then traced the issue to a re-allocation of the ION_STRING defined on line 49:
```C
ION_STRING istring;
```
This structure unfortunately isn't initialized. Within the structure is a pointer that is not initialized to NULL, and later the garbage address is free'd during a re-allocation to increase the string size.

I've fixed this locally by initializing the struct at the declaration using:
```C
ION_STRING istring = {0};
```

With this fix, the code runs as expected:
```
$ src/IonCInt 
ion_decimal: 1.23456
     double: 1234500.000000
        int: 123456
    ion_int: 12345678901234567890
```

I'll fix this in the cookbook and take a pass at the rest of the ion-c entries to make sure there aren't any other issues.

Since you mentioned that you were having issues prior to using the cookbook entry, is it possible that similarly uninitialized ION_STRINGs caused the issue you were experiencing?

**Comment by zabele** [comment]
Almost certainly, since i had basically copied the code more or less verbatim.I just added the initialization, and magically everything is now working correctly.Well done! And thanks a bunch for the timely response 😊SteveSent from my iPhoneOn Jun 27, 2023, at 6:05 PM, Richard Giliam ***@***.***> wrote:﻿
Hello zabele! Thank you for posting this issue, it looks like our cookbook entry has a bug.
I first duplicated a segmentation fault with the cookbook code:
$ src/IonCInt
ion_decimal: 1.23456
     double: 1234500.000000
        int: 123456
fish: Job 1, 'src/IonCInt' terminated by signal SIGSEGV (Address boundary error)

Then traced the issue to a re-allocation of the ION_STRING defined on line 49:
ION_STRING istring;
This structure unfortunately isn't initialized. Within the structure is a pointer that is not initialized to NULL, and later the garbage address is free'd during a re-allocation to increase the string size.
I've fixed this locally by initializing the struct at the declaration using:
ION_STRING istring = {0};
With this fix, the code runs as expected:
$ src/IonCInt 
ion_decimal: 1.23456
     double: 1234500.000000
        int: 123456
    ion_int: 12345678901234567890

I'll fix this in the cookbook and take a pass at the rest of the ion-c entries to make sure there aren't any other issues.
Since you mentioned that you were having issues prior to using the cookbook entry, is it possible that similarly uninitialized ION_STRINGs caused the issue you were experiencing?
